Wikipedia/Code/Article as a Living Document/ArticleAsLivingDocViewController.swift

```diff
@@ -0,0 +1,374 @@
+            return estimate
+        }
+        
+        sectionHeaderView.configure(viewModel: section, theme: theme)
+        
+        estimate.height = sectionHeaderView.sizeThatFits(CGSize(width: columnWidth, height: UIView.noIntrinsicMetric), apply: false).height
+        estimate.precalculated = true
+        return estimate
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, estimatedHeightForItemAt indexPath: IndexPath, forColumnWidth columnWidth: CGFloat) -> ColumnarCollectionViewLayoutHeightEstimate {
+        var estimate = ColumnarCollectionViewLayoutHeightEstimate(precalculated: false, height: 350)
+        
+        guard let event = dataSource.itemIdentifier(for: indexPath) else {
+            return estimate
+        }
+        
+        let cell: CollectionViewCell
+        switch event {
+        case .large(let largeEvent):
+            guard let largeEventCell = layoutManager.placeholder(forCellWithReuseIdentifier: ArticleAsLivingDocLargeEventCollectionViewCell.identifier) as? ArticleAsLivingDocLargeEventCollectionViewCell else {
+                return estimate
+            }
+            
+            
+            largeEventCell.configure(with: largeEvent, theme: theme)
+            cell = largeEventCell
+        case .small(let smallEvent):
+            guard let smallEventCell = layoutManager.placeholder(forCellWithReuseIdentifier: ArticleAsLivingDocSmallEventCollectionViewCell.identifier) as? ArticleAsLivingDocSmallEventCollectionViewCell else {
+                return estimate
+            }
+            
+            smallEventCell.configure(viewModel: smallEvent, theme: theme)
+            cell = smallEventCell
+        }
+        
+        cell.layoutMargins = layout.itemLayoutMargins
+        estimate.height = cell.sizeThatFits(CGSize(width: columnWidth, height: UIView.noIntrinsicMetric), apply: false).height
+        estimate.precalculated = true
+        
+        return estimate
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        
+        guard let articleAsLivingDocViewModel = delegate?.articleAsLivingDocViewModel else {
+            return
+        }
+
+        (cell as? ArticleAsLivingDocLargeEventCollectionViewCell)?.delegate = self
+        (cell as? ArticleAsLivingDocLargeEventCollectionViewCell)?.articleDelegate = self
+        (cell as? ArticleAsLivingDocSmallEventCollectionViewCell)?.delegate = self
+        
+        let numSections = dataSource.numberOfSections(in: collectionView)
+        let numEvents = dataSource.collectionView(collectionView, numberOfItemsInSection: indexPath.section)
+        
+        if indexPath.section == numSections - 1 &&
```

Comment:
It might be worth triggering this earlier than the last cell since the paging call is so slow. That can be an optimization that we do later though.